bkbit/scripts/gfftranslator.py

```diff
@@ -1,203 +1,682 @@
+                        # TODO: Write cleaner code that calls respective generate function based on the authority automatically
+                        if (
+                            self.genome_annotation.authority
+                            == kbmodel.AuthorityType.ENSEMBL
+                        ):
+                            gene_annotation = self.generate_ensembl_gene_annotation(
+                                attributes, curr_line_num
+                            )
+                            if gene_annotation is not None:
+                                self.gene_annotations[gene_annotation] = gene_annotation
+                        elif (
+                            self.genome_annotation.authority
+                            == kbmodel.AuthorityType.NCBI
+                        ):
+                            gene_annotation = self.generate_ncbi_gene_annotation(
+                                attributes, curr_line_num
+                            )
+                            if gene_annotation is not None:
+                                self.gene_annotations[gene_annotation.id] = (
+                                    gene_annotation
+                                )
+                progress_bar.update(1)
+                curr_line_num += 1
+            progress_bar.close()
+
+    def generate_ensembl_gene_annotation(self, attributes, curr_line_num):
+        """
+        Generates a GeneAnnotation object for Ensembl based on the provided attributes.
+
+        Args:
+            attributes (dict): A dictionary containing the attributes of the gene.
+            curr_line_num (int): The line number of the current row in the input file.
+
+        Returns:
+            GeneAnnotation or None: The generated GeneAnnotation object if it is not a duplicate,
+            otherwise None.
+
+        Raises:
+            None
+
+        """
+        stable_id = self.__get_attribute(attributes, "gene_id", curr_line_num)
+        if stable_id:
+            stable_id = stable_id.split(".")[0]
+
+        # Check and validate the name attribute
+        name = self.__get_attribute(attributes, "Name", curr_line_num)
+
+        # Check and validate the description attribute
+        description = self.__get_attribute(attributes, "description", curr_line_num)
+
+        # Check and validate the biotype attribute
+        biotype = self.__get_attribute(attributes, "biotype", curr_line_num)
+
+        gene_annotation = kbmodel.GeneAnnotation(
+            id=ENSEMBL_GENE_ID_PREFIX + ":" + stable_id,
+            source_id=stable_id,
+            symbol=name,
+            name=name,
+            description=description,
+            molecular_type=biotype,
+            referenced_in=self.genome_annotation.id,
+            in_taxon=[self.organism_taxon.id],
+            in_taxon_label=self.organism_taxon.full_name,
+        )
+        # handle duplicates
+        if gene_annotation not in self.gene_annotations:
```

Comment:
Something different may have to happen here, depending on you response to my comment at line 370

Reply:
My answer above should explain this line.